AI #1793: Create Commitment Program Eligibility Details column (#1882)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/commitmentprogrameligibility.md

> # Commitment Eligibility Details

Reviewer: Everyone’s favorite topic — naming 😅

While the old `CommitmentDiscountType` property name (now resolved) was genuinely problematic for me, here I’m not entirely sure what the better option is.

The name `CommitmentEligibilityDetails` was interpreted by one participant as referring to *Commitment Discount* eligibility. Given that we now have the *Commitment Program* glossary term, and this column actually refers to *Commitment Program* eligibility, I was wondering if we should consider renaming it to: `CommitmentProgramEligibilityDetails`.

**Pro:**
- Makes it explicit that this refers to *Commitment Program* eligibility (and not *Commitment Discount*), avoiding ambiguity.

**Con:**
- The name becomes quite long, and it would require additional changes 😢. A shorter, more practical name may initially be misleading, but could become intuitive over time.

Curious to hear your thoughts — is the added clarity worth the trade-off?

*P.S. As I’m writing this, I’m leaning towards keeping the current name (it will become intuitive/clear over time), but I’m raising this since we agreed to mention it.*

Reviewer 2: I keep thinking about @udam-stitcher's point in TF-3 yesterday that `Commitment` is not making clear that this is neither specifically `ContractCommitments` nor `CommitmentDiscounts`.  If we feel strongly that we need to implement this third `CommitmentProgram` construct, then we should have sufficient confidence to align the column name with the construct.

I am in favor of changing the name.  I agree that `CommitmentProgramEligibilityDetails` is pretty long 😅, though I personally think self-documentation always trumps brevity when naming things in a data model.

I would therefore alternatively suggest naming it `CommitmentProgramEligibility` (no `Details`).  This would follow the naming convention we used for a similar complex JSON column in ContractCommitment: `ContractCommitmentApplicability`.

Reviewer 3: On renaming to CommitmentEligibility (dropping "Details")

The "Details" suffix convention exists on columns like AllocatedMethodDetails and SkuPriceDetails where the JSON structure provides supplementary detail about something already represented by a parent/child column. Those columns answer "tell me more about X." CommitmentEligibility is different — the JSON structure is the answer, not additional detail about it. The column answers a single, direct question: "what commitment programs is this usage eligible for?" Since the column is in Cost and Usage dataset, I dont see a need to specifically call it commitmentdiscounteligibility or try and differentiate it from contract commitment columns because its not in the contract commitment dataset. If we had to have one for contracts we could simply call that ContractCommitmentEligibility.

Reviewer 4: If there's no strict limit on length I say we just go with the long name. I think suffixes like Type, Category, and Details have some value that we shouldn't make it inconsistent (even if we have in some cases already-that should just mean we fix those eventually)

Author: I agree with everything that's being said. I was never married to "Details" we put that in the name just for consistency with other JSONic columns. Commitment Programs does a good job highlighting that this spans across different programs like contracted and negotiated.

Reviewer 5: We discussed this again in TF-3 and we believe that CommitmentProgramEligibilityDetails, while verbose, is the most self-evident way to name this column (and it's only 35 characters out of the 50 character limit).

> Commitment Eligibility Details identifies the [*commitment programs*](#glossary:commitment-program) that could potentially [*cover charges*](#glossary:covered-charge), subject to [*service provider*](#glossary:service-provider) constraints. By distinguishing the pool of spend that was eligible to be covered, Commitment Eligibility Details provides the fundamental denominator for calculating precise commitment coverage metrics. This allows FinOps practitioners to accurately size the pool of uncovered spend that could realistically be covered by a future commitment. In this context, *commitment programs* include both discount-bearing constructs (e.g., Savings Plans, committed-use discounts) and advance capacity reservations (e.g., zonal reservations), provided the service provider treats them as [*commitments*](#glossary:commitment).

Reviewer 6: ```suggestion
Commitment Eligibility Details identifies the [*commitment programs*](#glossary:commitment-program) that could potentially [*cover charges*](#glossary:covered-charge), subject to [*service provider*](#glossary:service provider) constraints. By distinguishing the pool of spend that was eligible to be covered, Commitment Eligibility Details provides the fundamental denominator for calculating precise commitment coverage metrics. This allows FinOps practitioners to accurately size the pool of uncovered spend that could realistically be covered by a future commitment. In this context, *commitment programs* include both discount-bearing constructs (e.g., Savings Plans, committed-use discounts) and advance capacity reservations (e.g., zonal reservations), provided the service provider treats them as [*commitments*](#glossary:commitment).
```

Also, (#glossary:covered-charge) seems to be missing in the glossary.

Reviewer 7: 🤖 [AI][Claude] Heads up: this suggestion changes `#glossary:service-provider` (hyphenated) to `#glossary:service provider` (with a space). The glossary anchor uses the hyphenated form, so accepting this as-is would break the link.

Reviewer 2: @narlaadbe 
* As @flanakin points out, the convention is indeed `service-provider`, not `service provider`.
* Your point about `covered-charge` missing from this PR is fair.  It is being proposed via PR #1966, which is also being stewarded by TF-3.  We would depend on that PR to be approved, and then this link will resolve.

I believe no action is required here.

Reviewer 6: The glossary anchor for this particular entry is missing the hyphen. We should probably fix that.

https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/glossary.md?plain=1#L217

Author: Makes sense. Suggestion accepted.

## Requirements

CommitmentEligibilityDetails MUST adhere to the following requirements:

* CommitmentEligibilityDetails MUST be of type String.
* CommitmentEligibilityDetails MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* CommitmentEligibilityDetails MUST conform to [JsonObjectFormat](#attributes.jsonobjectformat) requirements.
* CommitmentEligibilityDetails MUST NOT be null when a charge is eligible for a [*commitment program*](#glossary:commitment-program), regardless of whether a [*commitment*](#glossary:commitment) was actually applied to the charge.
* CommitmentEligibilityDetails MUST NOT consider transient account configurations or quotas that might temporarily prevent purchase or participation in a *commitment program*.
* CommitmentEligibilityDetails MUST include all publicly available *commitment programs* for which the usage is eligible.

> * CommitmentEligibilityDetails SHOULD include negotiated *commitment programs* for which the usage is eligible.

Reviewer 7: To be clear, is this saying that data gens should add this to EVERY row where negotiated discounts are eligible? That's a lot! Microsoft has this as a Boolean, but the amount of extra ceremony in this column makes this fairly big. Not pushing back, just confirming I'm reading this correctly and calling it out. I'm okay either way.

Reviewer 2: Discussed this in TF-3.  No consensus, but:

@flanakin believed it's okay as-is -- just calling it out.
@udam-stitcher suggested perhaps making a more explicit declaration here to _exclude_ such programs.  (If he feels this is necessary, he will leave a suggestion.)

Lacking a more explicit suggestion, we would take no action and leave as-is.

Reviewer 3: I'd recommend removing this line. Negotiated commitments are contractual arrangements between a provider and customer — they fall under contract commitments, not commitment programs. For example, a private rate discount for specific services is contracted and negotiated; it doesn't align to the concept of a publicly available commitment program like a Savings Plan or Reserved Instance. Mixing the two here muddies the scope of this column, which should focus on what a practitioner can go out and purchase today. Negotiated arrangements are better served by the Contract Commitment dataset.

Author: Let me provide some context for why this line is included. We are trying to make accommodations for things like Oracle's Universal credits https://www.oracle.com/cloud/universal-credits/ or Google's SUDs https://docs.cloud.google.com/compute/docs/sustained-use-discounts. These are publicly referenced programs but do not fall under the typical spend or resource-based commitment types. SaaS providers also have some hybrid programs that offer a discount and are publicly referenced, but do not fit the typical commitment rules. 
This is optional (SHOULD and not MUST). A provider can decide what they want to list in here. If it is not public, we don't expect to see it in here. If it is public and the provider deems it a commitment program, then they are welcome to add it in here. 
I am not expecting to see any negotiated programs that we're not allowed to talk about.

Reviewer 5: TF-3 addressed this on April 3. The group changed requirement 15 from SHOULD to MAY for negotiated commitment programs, with language narrowing scope: providers MAY include negotiated programs when the usage is eligible and the program is not broadly applicable across the service provider's service catalog. This gives providers discretion to avoid the exact scenario you flagged (populating every row with a globally applicable program). See the pending suggestion on requirement 15.

> * CommitmentEligibilityDetails SHOULD include negotiated *commitment programs* for which the usage is eligible.

Reviewer 7: How do we expect this to work for vendors who bill based on managed usage/spend and may offer discounts on all rows or maybe even just a subset of rows? Would they need to add differentiated programs separate from the ones that represent the provider's? So in theory, this would be layered? 
- Vendor negotiated discount for managed services
- SaaS provider negotiated discount (via cloud marketplace)
- Cloud provider negotiated discount

I'm wondering if this is clear about how to handle that 🤔 Again, not pushing back. Just thinking out loud. I'm curious how others might interpret it. Not specifically this line, but probably more holistically.

Author: That's an interesting point. I don't think we considered MSP when we wrote these. 
My opinion is that, if a provider gives this option - even if it is managed, it is still doing its job, exposing all possible options of commitments for a row, irrespective of who is going to be clicking the buttons and making the commitments.

Reviewer 5: TF-3 discussed MSP handling on April 3. The group agreed this is a broader FOCUS gap that isn't specific to this column. Sanjna's framing holds for now: the column exposes commitment programs from the billing data generator's perspective, regardless of who manages the purchasing relationship. If MSP-specific guidance is needed, it should be addressed across the spec rather than scoped to this PR.

* CommitmentEligibilityDetails MUST NOT include data related to *commitment* [*periods*](#glossary:period) or payment options.
* CommitmentEligibilityDetails MUST conform to [CommitmentEligibilityDetailsObject](#datasets.costandusage.commitmenteligibilitydetails.commitmenteligibilitydetailsobject) requirements when CommitmentEligibilityDetails is not null.

## Commitment Eligibility Details Object

Commitment Eligibility Details consists of a valid JSON object with a top-level property key `CommitmentPrograms` containing an array of objects describing the specific [*commitment programs*](#glossary:commitment-program) available for the usage charge.

### Object Requirements

CommitmentEligibilityDetailsObject MUST adhere to the following requirements:

* CommitmentEligibilityDetailsObject MUST have a top-level property key "CommitmentPrograms".
* CommitmentEligibilityDetailsObject MAY contain additional data generator-defined top-level property keys.
* CommitmentEligibilityDetailsObject MUST have property keys that begin with the string "x\_" unless it is a FOCUS-defined property key.

> * CommitmentEligibilityDetailsObject.CommitmentPrograms MUST adhere to the following requirements:

Reviewer 7: 🤖 [AI][Claude] The `x\_` prefix uses a backslash escape in the markdown source (here and on the nested requirement below). Other JSON columns (AllocatedMethodDetails, ContractApplied) use `x_` without the escape. Both render identically, but removing the backslash would align with the established source pattern.

  * CommitmentEligibilityDetailsObject.CommitmentPrograms MUST be of type Array.

>   * CommitmentEligibilityDetailsObject.CommitmentPrograms MUST contain one or more objects.

Reviewer 7: Fwiw, I've had BI engineers ask to default to `{}` rather than null for JSON columns. Given the nesting here, I wouldn't be surprised if some would ask for an empty array (`{ CommitmentPrograms: [] }`). Should we allow some affordance for that? I don't have a preference. I get the usability benefit, but also get the value of null when there's no value. I'd defer to BI engineers to really comment on this. @shawnalpay?

Reviewer 7: In TF3, we agreed we should stick with null for consistency. I'm good with that.

  * Each entry in CommitmentEligibilityDetailsObject.CommitmentPrograms MUST be of type JSON Object.
  * CommitmentEligibilityDetailsObject.CommitmentPrograms MUST have a property key "ProgramType".
  * CommitmentEligibilityDetailsObject.CommitmentPrograms MAY contain additional data generator-defined property keys.

>   * CommitmentEligibilityDetailsObject.CommitmentPrograms MUST have property keys that begin with the string "x\_" unless it is a FOCUS-defined property key.

Reviewer 6: Minor: Since CommitmentPrograms is an array and not an Object
```suggestion
  * Each entry in CommitmentEligibilityDetailsObject.CommitmentPrograms MUST have a property key "ProgramType".
  * Each entry in CommitmentEligibilityDetailsObject.CommitmentPrograms MAY contain additional data generator-defined property keys.
  * Each entry in CommitmentEligibilityDetailsObject.CommitmentPrograms MUST have property keys that begin with the string "x\_" unless it is a FOCUS-defined property key.

```

  * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST be of type String.
  * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST NOT be null.

>   * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST correspond to a *commitment program* type supported by the service provider (e.g., "SavingsPlan", "ReservedInstance", "CommittedUseDiscount", "CapacityReservation", "ZonalReservation").

Reviewer 7: FOCUS standardizes on display strings. This is generally mentioned in descriptions. For instance, `Contract Commitment Type is a readable display name and not a code.` This is effectively saying it should have spaces and not be an enumeration.
```suggestion
  * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST correspond to a *commitment program* type supported by the service provider (e.g., "Savings Plan", "Reserved Instance", "Committed Use Discount", "Capacity Reservation", "Zonal Reservation").
```

Reviewer 2: Discussed in TF-3.  While it would be okay to accept this suggestion, we currently do not have any normative statements requiring this.  

To be clear, @flanakin: if we accept this change, it still would not be going as far as requiring data generators to follow this convention.  If you think that's important, then you should suggest adding such a requirement.

Reviewer 7: While I get the the fact that we don't have a hard requirement for this, we should lead by example. If we keep it as-is, then this will be the ONLY place we use a "code" for a value. We should be consistent, even if it's not a hard requirement. 

As a note, if we don't fix this, we will actually be inconsistent with AWS and Microsoft. Both of them use spaces. I don't see sample data from GCP, so I can't comment on them or other providers. That alone sends the signal to them that they are doing it wrong, if we show an example without spaces. 

<img width="768" height="222" alt="image" src="https://github.com/user-attachments/assets/4edf196b-3eb6-47f4-a584-8f2c26dcb93d" />

[See focus.finops.org/sandbox/#uc-14403](https://focus.finops.org/sandbox/#uc-14403)

Reviewer 2: @flanakin 
I just chatted with @Matt-Cowsert about this.  **How would you feel about us cutting the examples from this normative requirement entirely?**  That doesn't change your overall concern (see the other thread for my response on that), but it would eliminate our need to make further changes to normative requirements to close on this.

>   * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST correspond to a *commitment program* type supported by the service provider.

Reviewer 2: Suggested a path forward in [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1882/files#r3055361467), given that this thread is outdated.

Reviewer 2: @SanjnaSrivatsaVMware accepted the suggestion I made in that other thread.  Resolving!

>   * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST equal [CommitmentDiscountType](#datasets.costandusage.commitmentdiscounttype) for one object in CommitmentEligibilityDetailsObject.CommitmentPrograms when CommitmentDiscountType is not null.

Reviewer 7: Let's make sure our data is consistent with itself even when it's not directly applied on the row 🙂 I added a new requirement that requires the program type to match the values in the CommitmentDiscountType column even if the current row does not have a commitment discount applied. 

So, if a savings plan row would use "Savings Plan" for CommtimentDiscountType, then ProgramType must use "Savings Plan" for any row where that is eligible, even if not applied. Happy to reword if that's not clear enough.

```suggestion
  * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST equal the value used in [CommitmentDiscountType](#datasets.costandusage.commitmentdiscounttype) when the *commitment program* is the same [*commitment discount*](#glossary:commitment-discount) type.
  * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST equal [CommitmentDiscountType](#datasets.costandusage.commitmentdiscounttype) for one object in CommitmentEligibilityDetailsObject.CommitmentPrograms when CommitmentDiscountType is not null.
```

Reviewer 7: Fwiw, after discussing in TF3, I realized that the existing requirement does lead to "eventual consistency" which might be adequate for our needs. Meaning, while we don't require all rows use the same values explicitly, assuming one row would match, that would require them to be globally consistent.

This reminds me of SkuPriceDetails, where we talk about requiring consistency across rows. That's slightly different, but the same concept.

Reviewer 2: Agree that we should take no action here. 

That said, @flanakin, I think the following reword would more clearly sell the point you were making in TF-3, but which I was really struggling to grok based on what you originally wrote.  Going back and forth with Gemini produces the following:

> * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST equal the [CommitmentDiscountType](#datasets.costandusage.commitmentdiscounttype) value associated with that *commitment program*, even if CommitmentDiscountType is null on the current row (e.g., use "Savings Plan" in both, not "SP" in one and "Savings Plan" in the other).

Reviewer: Instead of enforcing terminology consistency via `CommitmentEligibilityDetails`, I think this should be defined on the `CommitmentDiscountType` column itself, rather than on a column that references it.

We already use similar wording elsewhere to express consistency expectations over time, for example:
- `SkuPriceId MUST remain consistent over time.`
- `Existing SkuPriceDetails properties SHOULD remain consistent over time.`

Following that pattern, I think a `SHOULD` requirement would be more appropriate here. While consistency is clearly desirable, display names (unlike IDs) don’t necessarily need to be strictly immutable.

---

This is probably **out of scope for this PR**, but as a follow-up I’d suggest adding something like this to `CommitmentDiscountType`:

> - `CommitmentDiscountType SHOULD remain consistent over time for a given service provider.`
> - `CommitmentDiscountType SHOULD correspond to terminology disclosed by the service provider in public documentation.`

This would make the expectation explicit at the source.

Reviewer: Assuming this is out-of-scope of this PR, I just created the Issue #2119

Author: Thank you folks! No action will be taken here. Ignoring suggestion

  * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType SHOULD correspond to terminology disclosed by the service provider in public documentation.

### Top-Level Properties

| Property             | Type  | Required | Description                                                                         |
|:----------|:----------|:----------|:---------------------------------------|
| `CommitmentPrograms` | Array | True     | Array of objects identifying *commitment programs* for which the usage is eligible. |

### Example Entries

> | Key         | ValueType                            | Required | Description                                                                                                |

Reviewer 7: 🤖 [AI][Claude] Minor: This table uses "ValueType" as a column header. Other JSON columns (AllocatedMethodDetails, ContractApplied) use "Type" consistently. Consider aligning for pattern consistency.
```suggestion
| Key         | Type                                 | Required | Description                                                                                                |
```

Reviewer 2: Discussed in Mar 27 TF-3.  I have action item #2117 to tackle this if/when this PR is approved.

No action here.

Author: Ack. Not accepting the suggestion

> |:-------------|:-------------|:-------------|:------------------------------|

Reviewer 7: We need something that can be generalized across providers for this to scale. I'd really like to see a ProgramCategory that can classify "Commitment Discount", "Negotiated Discount", "Capacity Reservation", "Other". I do understand we can add this later. I'm not going to block on this. I'm open to whichever way the community wants to go.
```suggestion
|:-------------|:-------------|:-------------|:------------------------------|
| ProgramCategory | [String](#attributes.stringhandling) | True     | Provider-agnostic classification of *commitment program*. |
```

| ProgramType | [String](#attributes.stringhandling) | True     | The specific type of commitment program (e.g., discount or capacity reservation) available for this usage. |

### Object Example

``` json
{
  "CommitmentPrograms": [
    { "ProgramType": "SavingsPlan" },
    { "ProgramType": "ReservedInstance" },
    { "ProgramType": "CapacityReservation" },

>     { "ProgramType": "ZonalReservation" }

Reviewer 7: Aligning the example to our general approach of using display strings and not codes (enumerations).
```suggestion
    { "ProgramType": "Savings Plan" },
    { "ProgramType": "Reserved Instance" },
    { "ProgramType": "Capacity Reservation" },
    { "ProgramType": "Zonal Reservation" }
```

Reviewer 2: Discussed in TF-3. While it would be okay to accept this suggestion, we currently do not have any normative statements requiring this.

To be clear, @flanakin: if we accept this change, it still would not be going as far as requiring data generators to follow this convention. If you think that's important, then you should suggest adding such a requirement.

Author: Thank you. Accepting suggestion

Reviewer 7: Unresolving for the same reason and justification as before. AWS and Azure use spaces and using "codes" here sends the message that they are doing it wrong. Since we clearly modeled this after AWS (based on the names being used), I would argue this should be more realistic with what's in the real world. Otherwise, it opens confusion.

Reviewer 2: @flanakin 
If you agree with the proposal I made in the [other thread you opened](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1882#discussion_r3054809097), would you be comfortable with us logging an AI to tackle this change during consistency review?  

There are a non-trivial number of places that @SanjnaSrivatsaVMware and @Matt-Cowsert would have to change in both #1882 and #2025, and we're short on time, but if it's an informative change, then that takes the pressure off.

@Matt-Cowsert is okay with making this change (though, as I said above, the change itself would not enforce anything -- it would be purely informative).  Not sure if @SanjnaSrivatsaVMware is onboard.

Author: I'm onboard too. Good to tackle this during consistency review. I don't want to attempt to change these today.

Reviewer 5: I've create #2178 to capture the scope of this ask. Can tweak if needed.

Reviewer 2: Resolving per the above agreement of @SanjnaSrivatsaVMware and @Matt-Cowsert to tackle this via AI #2178 in consistency review.  @flanakin please reopen if you believe this to be insufficient.

### Provider C (Ineligible Object Storage Usage)

Scenario: Standard object storage usage or a support fee, which is not eligible for any commitment program.

| ServiceProviderName | ServiceName   | CommitmentEligibilityDetails |
|---------------------|---------------|------------------------------|
| Provider C          | ObjectStorage | null                         |

### Provider D (Capacity Reservation-eligible Compute Usage)

A compute instance type and tenancy that are eligible for both Savings Plans/Reserved Instances and for capacity reservations (e.g., regional reservations, zonal reservations). The eligibility column reflects all commitment constructs the usage qualifies for.

| ServiceProviderName | ServiceName    | CommitmentEligibilityDetails                                                                                                                                               |
|------------------|------------------|------------------|
| Provider D          | Compute | {"CommitmentPrograms": [{"ProgramType": "SavingsPlan"}, {"ProgramType": "ReservedInstance"}, {"ProgramType": "CapacityReservation"}, {"ProgramType": "ZonalReservation"}]} |

> ### Object ID

Reviewer 7: 🤖 [AI][Claude] "Object ID" and "Object Display Name" (### headings) appear nested under "## Example Scenarios" due to heading hierarchy. In other JSON columns (AllocatedMethodDetails, ContractApplied), these live under a "## Implementation Guidance" parent section. Consider adding that parent heading before these, or moving them outside Example Scenarios.

Reviewer 2: Discussed in Mar 27 TF-3.  I have action item #2117 to tackle this if/when this PR is approved.

No action here.

Author: Ack. No action here

CommitmentEligibilityDetailsObject

### Object Display Name

Commitment Eligibility Details Object

## Column ID

CommitmentEligibilityDetails

## Display Name

Commitment Eligibility Details

## Description

The types of [*commitment programs*](#glossary:commitment-program) available for a specific usage row.

> ## Content constraints

Reviewer 7: 🤖 [AI][Claude] Capitalization: "Content constraints" should be "Content Constraints" for consistency with other column definitions (e.g., AllocatedMethodDetails, ContractApplied).
```suggestion
## Content Constraints
```

# Commitment Program Eligibility

> Commitment Program Eligibility identifies the [*commitment programs*](#glossary:commitment-program) that could potentially cover [*charges*](#glossary:charge), subject to [*service provider*](#glossary:service-provider) constraints. By distinguishing the pool of spend that was eligible to be covered, Commitment Program Eligibility provides the fundamental denominator for calculating precise commitment coverage metrics. This allows FinOps practitioners to accurately size the pool of uncovered spend that could realistically be covered by a future commitment. In this context, *commitment programs* include both discount-bearing programs (e.g., Flexible Spend Plans, Resource Reservations) and advance capacity reservations (e.g., Advance Capacity Reservations), provided the service provider treats them as [*commitments*](#glossary:commitment).

Reviewer 5: Align narrative with PR #2115 Fictitious Commitment Program Reference: "Advance Capacity Reservations" → "Advance Resource Commitments (ARCs)"

```suggestion
Commitment Program Eligibility identifies the [*commitment programs*](#glossary:commitment-program) that could potentially cover [*charges*](#glossary:charge), subject to [*service provider*](#glossary:service-provider) constraints. By distinguishing the pool of spend that was eligible to be covered, Commitment Program Eligibility provides the fundamental denominator for calculating precise commitment coverage metrics. This allows FinOps practitioners to accurately size the pool of uncovered spend that could realistically be covered by a future commitment. In this context, *commitment programs* include both discount-bearing programs (e.g., Flexible Spend Plans, Resource Reservations) and advance resource commitments (e.g., Advance Resource Commitments), provided the service provider treats them as [*commitments*](#glossary:commitment).
```

* CommitmentProgramEligibilityObject MUST have property keys that begin with the string "x_" unless it is a FOCUS-defined property key.
* CommitmentProgramEligibilityObject.CommitmentPrograms MUST adhere to the following requirements:
  * CommitmentProgramEligibilityObject.CommitmentPrograms MUST be of type Array.
  * CommitmentProgramEligibilityObject.CommitmentPrograms MUST contain one or more objects.
  * Each entry in CommitmentProgramEligibilityObject.CommitmentPrograms MUST be of type JSON Object.
  * Each entry in CommitmentProgramEligibilityObject.CommitmentPrograms MUST have a property key "ProgramType".
  * Each entry in CommitmentProgramEligibilityObject.CommitmentPrograms MAY contain additional data generator-defined property keys.
  * Each entry in CommitmentProgramEligibilityObject.CommitmentPrograms MUST have property keys that begin with the string "x_" unless it is a FOCUS-defined property key.
  * CommitmentProgramEligibilityObject.CommitmentPrograms.ProgramType MUST be of type String.
  * CommitmentProgramEligibilityObject.CommitmentPrograms.ProgramType MUST NOT be null.

>   * CommitmentProgramEligibilityObject.CommitmentPrograms.ProgramType MUST correspond to a *commitment program* type supported by the service provider (e.g., "FlexibleSpendPlan", "ResourceReservation", "CapacityCreditPool", "AdvanceCapacityReservation").

Reviewer 5: Align ProgramType examples with PR #2115: "CapacityCreditPool" → "BulkCapacityCredit" (BCCs), "AdvanceCapacityReservation" → "AdvanceResourceCommitment" (ARCs)

```suggestion
  * CommitmentProgramEligibilityObject.CommitmentPrograms.ProgramType MUST correspond to a *commitment program* type supported by the service provider (e.g., "FlexibleSpendPlan", "ResourceReservation", "BulkCapacityCredit", "AdvanceResourceCommitment").
```

### Top-Level Properties

| Property             | Type  | Required | Description                                                                         |
|:----------|:----------|:----------|:---------------------------------------|
| `CommitmentPrograms` | Array | True     | Array of objects identifying *commitment programs* for which the usage is eligible. |

### Example Entries

| Key         | ValueType                            | Required | Description                                                                                                |
|:-------------|:-------------|:-------------|:------------------------------|
| ProgramCategory | [String](#attributes.stringhandling) | True     | Provider-agnostic classification of *commitment program*. |
| ProgramType | [String](#attributes.stringhandling) | True     | The specific type of commitment program (e.g., discount or capacity reservation) available for this usage. |

### Object Example

``` json
{
  "CommitmentPrograms": [
    { "ProgramType": "FlexibleSpendPlan" },
    { "ProgramType": "ResourceReservation" },

>     { "ProgramType": "AdvanceCapacityReservation" },

Reviewer 5: ARC alignment (1/2): "AdvanceCapacityReservation" → "AdvanceResourceCommitment" per PR #2115

```suggestion
    { "ProgramType": "AdvanceResourceCommitment" },
```

>     { "ProgramType": "ZonalCapacityReservation" }

Reviewer 5: ARC alignment (2/2): "ZonalCapacityReservation" → "ZonalResourceCommitment" per PR #2115

```suggestion
    { "ProgramType": "ZonalResourceCommitment" }
```

Author: I like how "out of the box" these new names are, haha!

## Example Scenarios

The examples below are not exhaustive and may change over time. Service providers are the authoritative source for their [*commitment programs*](#glossary:commitment-program).

### Aura Web (Partially Covered Compute Usage)

Scenario: A compute usage row that is partially covered by a Flexible Spend Plan. The eligibility column still reflects all programs this usage qualifies for, regardless of current coverage.

| ServiceProviderName | ServiceName | CommitmentProgramEligibility                                                                            |
|---------------------|-------------|---------------------------------------------------------------------------------------------------------|
| Aura Web            | Compute     | {"CommitmentPrograms": [{"ProgramType": "FlexibleSpendPlan"}, {"ProgramType": "ResourceReservation"}]} |

> ### StackLens (Observability with Platform Commitment)

Reviewer 5: ISC alignment (1/3): Update heading to reference Interval Spend Commitments per PR #2115

```suggestion
### StackLens (Observability with Interval Spend Commitment)
```

> Scenario: An observability platform usage row eligible for Monthly and Annual commitment-based pricing, offering lower effective rates than standard usage.

Reviewer 5: ISC alignment (2/3): Update scenario description to reference interval spend commitment pricing per PR #2115

```suggestion
Scenario: An observability platform usage row eligible for Monthly and Annual interval spend commitment pricing, offering lower effective rates than standard usage.
```

| ServiceProviderName | ServiceName   | CommitmentProgramEligibility                                                                                            |
|---------------------|---------------|-------------------------------------------------------------------------------------------------------------------------|

> | StackLens           | Observability | {"CommitmentPrograms": [{"ProgramType": "MonthlyPlatformCommitment"}, {"ProgramType": "AnnualPlatformCommitment"}]} |

Reviewer 5: ISC alignment (3/3): "MonthlyPlatformCommitment" → "MonthlyIntervalSpendCommitment", "AnnualPlatformCommitment" → "AnnualIntervalSpendCommitment" per PR #2115

```suggestion
| StackLens           | Observability | {"CommitmentPrograms": [{"ProgramType": "MonthlyIntervalSpendCommitment"}, {"ProgramType": "AnnualIntervalSpendCommitment"}]} |
```

### LatticeScale (Ineligible Object Storage Usage)

Scenario: Standard object storage usage or a support fee, which is not eligible for any commitment program.

| ServiceProviderName | ServiceName   | CommitmentProgramEligibility |
|---------------------|---------------|------------------------------|
| LatticeScale        | ObjectStorage | null                         |

> ### Aura Web (Capacity Reservation-Eligible Compute Usage)

Reviewer 5: ARC alignment (1/3): Update heading to reference Advance Resource Commitments per PR #2115

```suggestion
### Aura Web (Advance Resource Commitment-Eligible Compute Usage)
```

> Scenario: A compute instance type and tenancy that are eligible for both discount-bearing programs and capacity reservations. The eligibility column reflects all commitment constructs the usage qualifies for.

Reviewer 5: ARC alignment (2/3): Update scenario description to reference advance resource commitments per PR #2115

```suggestion
Scenario: A compute instance type and tenancy that are eligible for both discount-bearing programs and advance resource commitments. The eligibility column reflects all commitment constructs the usage qualifies for.
```

| ServiceProviderName | ServiceName | CommitmentProgramEligibility                                                                                                                                                                           |
|---------------------|-------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | Aura Web            | Compute     | {"CommitmentPrograms": [{"ProgramType": "FlexibleSpendPlan"}, {"ProgramType": "ResourceReservation"}, {"ProgramType": "AdvanceCapacityReservation"}, {"ProgramType": "ZonalCapacityReservation"}]} |

Reviewer 5: ARC alignment (3/3): ProgramType values updated to match PR #2115 ARCs naming

```suggestion
| Aura Web            | Compute     | {"CommitmentPrograms": [{"ProgramType": "FlexibleSpendPlan"}, {"ProgramType": "ResourceReservation"}, {"ProgramType": "AdvanceResourceCommitment"}, {"ProgramType": "ZonalResourceCommitment"}]} |
```

> # Commitment Eligibility Details

Reviewer 5: ```suggestion
# Commitment Program Eligibility
```

> Commitment Eligibility Details identifies the [*commitment programs*](#glossary:commitment-program) that could potentially [*cover charges*](#glossary:covered-charge), subject to [*service provider*](#glossary:service provider) constraints. By distinguishing the pool of spend that was eligible to be covered, Commitment Eligibility Details provides the fundamental denominator for calculating precise commitment coverage metrics. This allows FinOps practitioners to accurately size the pool of uncovered spend that could realistically be covered by a future commitment. In this context, *commitment programs* include both discount-bearing constructs (e.g., Savings Plans, committed-use discounts) and advance capacity reservations (e.g., zonal reservations), provided the service provider treats them as [*commitments*](#glossary:commitment).

Reviewer 5: ```suggestion
Commitment Program Eligibility identifies the [*commitment programs*](#glossary:commitment-program) that could potentially cover [*charges*](#glossary:charge), subject to [*service provider*](#glossary:service-provider) constraints. By distinguishing the pool of spend that was eligible to be covered, Commitment Program Eligibility provides the fundamental denominator for calculating precise commitment coverage metrics. This allows FinOps practitioners to accurately size the pool of uncovered spend that could realistically be covered by a future commitment. In this context, *commitment programs* include both discount-bearing programs (e.g., Flexible Spend Plans, Resource Reservations) and advance capacity reservations (e.g., Advance Capacity Reservations), provided the service provider treats them as [*commitments*](#glossary:commitment).
```

Fixes: rename (x2), broken `service-provider` anchor, broken `covered-charge` link (replaced with `[*charges*](#glossary:charge)` until PR #1966 merges), genericized real-world examples.

## Requirements

CommitmentEligibilityDetails MUST adhere to the following requirements:

* CommitmentEligibilityDetails MUST be of type String.
* CommitmentEligibilityDetails MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* CommitmentEligibilityDetails MUST conform to [JsonObjectFormat](#attributes.jsonobjectformat) requirements.
* CommitmentEligibilityDetails MUST NOT be null when a charge is eligible for a [*commitment program*](#glossary:commitment-program), regardless of whether a [*commitment*](#glossary:commitment) was actually applied to the charge.
* CommitmentEligibilityDetails MUST NOT consider transient account configurations or quotas that might temporarily prevent purchase or participation in a *commitment program*.
* CommitmentEligibilityDetails MUST include all publicly available *commitment programs* for which the usage is eligible.
* CommitmentEligibilityDetails SHOULD include negotiated *commitment programs* for which the usage is eligible.
* CommitmentEligibilityDetails MUST NOT include data related to *commitment* [*periods*](#glossary:period) or payment options.

> * CommitmentEligibilityDetails MUST conform to [CommitmentEligibilityDetailsObject](#datasets.costandusage.commitmenteligibilitydetails.commitmenteligibilitydetailsobject) requirements when CommitmentEligibilityDetails is not null.

Reviewer 5: ```suggestion
CommitmentProgramEligibility MUST adhere to the following requirements:

* CommitmentProgramEligibility MUST be of type String.
* CommitmentProgramEligibility MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* CommitmentProgramEligibility MUST conform to [JsonObjectFormat](#attributes.jsonobjectformat) requirements.
* CommitmentProgramEligibility MUST NOT be null when a charge is eligible for a [*commitment program*](#glossary:commitment-program), regardless of whether a [*commitment*](#glossary:commitment) was actually applied to the charge.
* CommitmentProgramEligibility MUST NOT consider transient account configurations or quotas that might temporarily prevent purchase or participation in a *commitment program*.
* CommitmentProgramEligibility MUST include all publicly available *commitment programs* for which the usage is eligible.
* CommitmentProgramEligibility SHOULD include negotiated *commitment programs* for which the usage is eligible.
* CommitmentProgramEligibility MUST NOT include data related to *commitment* [*periods*](#glossary:period) or payment options.
* CommitmentProgramEligibility MUST conform to [CommitmentProgramEligibilityObject](#datasets.costandusage.commitmentprogrameligibility.commitmentprogrameligibilityobject) requirements when CommitmentProgramEligibility is not null.
```

> ## Commitment Eligibility Details Object

Reviewer 5: ```suggestion
## Commitment Program Eligibility Object
```

> Commitment Eligibility Details consists of a valid JSON object with a top-level property key `CommitmentPrograms` containing an array of objects describing the specific [*commitment programs*](#glossary:commitment-program) available for the usage charge.

Reviewer 5: ```suggestion
Commitment Program Eligibility consists of a valid JSON object with a top-level property key `CommitmentPrograms` containing an array of objects describing the specific [*commitment programs*](#glossary:commitment-program) available for the usage charge.
```

  * Each entry in CommitmentEligibilityDetailsObject.CommitmentPrograms MUST have a property key "ProgramType".
  * Each entry in CommitmentEligibilityDetailsObject.CommitmentPrograms MAY contain additional data generator-defined property keys.
  * Each entry in CommitmentEligibilityDetailsObject.CommitmentPrograms MUST have property keys that begin with the string "x\_" unless it is a FOCUS-defined property key.

  * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST be of type String.
  * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST NOT be null.
  * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST correspond to a *commitment program* type supported by the service provider (e.g., "Savings Plan", "Reserved Instance", "Committed Use Discount", "Capacity Reservation", "Zonal Reservation").
  * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST equal [CommitmentDiscountType](#datasets.costandusage.commitmentdiscounttype) for one object in CommitmentEligibilityDetailsObject.CommitmentPrograms when CommitmentDiscountType is not null.

>   * CommitmentEligibilityDetailsObject.CommitmentPrograms.ProgramType SHOULD correspond to terminology disclosed by the service provider in public documentation.

Reviewer 5: ```suggestion
CommitmentProgramEligibilityObject MUST adhere to the following requirements:

* CommitmentProgramEligibilityObject MUST have a top-level property key "CommitmentPrograms".
* CommitmentProgramEligibilityObject MAY contain additional data generator-defined top-level property keys.
* CommitmentProgramEligibilityObject MUST have property keys that begin with the string "x_" unless it is a FOCUS-defined property key.
* CommitmentProgramEligibilityObject.CommitmentPrograms MUST adhere to the following requirements:
  * CommitmentProgramEligibilityObject.CommitmentPrograms MUST be of type Array.
  * CommitmentProgramEligibilityObject.CommitmentPrograms MUST contain one or more objects.
  * Each entry in CommitmentProgramEligibilityObject.CommitmentPrograms MUST be of type JSON Object.
  * Each entry in CommitmentProgramEligibilityObject.CommitmentPrograms MUST have a property key "ProgramType".
  * Each entry in CommitmentProgramEligibilityObject.CommitmentPrograms MAY contain additional data generator-defined property keys.
  * Each entry in CommitmentProgramEligibilityObject.CommitmentPrograms MUST have property keys that begin with the string "x_" unless it is a FOCUS-defined property key.
  * CommitmentProgramEligibilityObject.CommitmentPrograms.ProgramType MUST be of type String.
  * CommitmentProgramEligibilityObject.CommitmentPrograms.ProgramType MUST NOT be null.
  * CommitmentProgramEligibilityObject.CommitmentPrograms.ProgramType MUST correspond to a *commitment program* type supported by the service provider (e.g., "FlexibleSpendPlan", "ResourceReservation", "CapacityCreditPool", "AdvanceCapacityReservation").
  * CommitmentProgramEligibilityObject.CommitmentPrograms.ProgramType MUST equal [CommitmentDiscountType](#datasets.costandusage.commitmentdiscounttype) for one object in CommitmentProgramEligibilityObject.CommitmentPrograms when CommitmentDiscountType is not null.
  * CommitmentProgramEligibilityObject.CommitmentPrograms.ProgramType SHOULD correspond to terminology disclosed by the service provider in public documentation.
```

Also fixes: `x\_` to `x_` (2 occurrences), removes blank line between sub-bullets (line 37), genericizes ProgramType examples.

Author: I had a commit made for this yesterday that failed for some reason, so I'll accept this suggestion to fastrack things. I will stop using R studio as my IDE for good. Github does NOT like it

### Top-Level Properties

| Property             | Type  | Required | Description                                                                         |
|:----------|:----------|:----------|:---------------------------------------|
| `CommitmentPrograms` | Array | True     | Array of objects identifying *commitment programs* for which the usage is eligible. |

### Example Entries

| Key         | ValueType                            | Required | Description                                                                                                |
|:-------------|:-------------|:-------------|:------------------------------|

> | ProgramCategory | [String](#attributes.stringhandling) | True     | Provider-agnostic classification of *commitment program*. |

Reviewer 2: What is happening here with `ProgramCategory`?  Are we truly carrying this in the definition along with `ProgramType`?  If so, we need to add it everywhere.  Otherwise, we need to strike it.

Reviewer 5: ProgramCategory is a leftover from a prior concept that distinguished category (e.g., "Discount" vs "CapacityReservation") from type (the specific program name). That got collapsed into just ProgramType during iteration, but ProgramCategory was not cleaned out of the Example Entries table.

It should be struck. It is not in the normative requirements, the Object Requirements, the JSON Type Definition, the example JSON object, or any example scenario. Only appears in this one table.

@SanjnaSrivatsaVMware — flagging so you can remove the ProgramCategory row from the Example Entries table when you are next in this file.

Author: Removed!

| ProgramType | [String](#attributes.stringhandling) | True     | The specific type of commitment program (e.g., discount or capacity reservation) available for this usage. |

### Object Example

``` json
{
  "CommitmentPrograms": [
    { "ProgramType": "Savings Plan" },
    { "ProgramType": "Reserved Instance" },
    { "ProgramType": "Capacity Reservation" },
    { "ProgramType": "Zonal Reservation" }

>   ]

Reviewer 5: ```suggestion
  "CommitmentPrograms": [
    { "ProgramType": "FlexibleSpendPlan" },
    { "ProgramType": "ResourceReservation" },
    { "ProgramType": "AdvanceCapacityReservation" },
    { "ProgramType": "ZonalCapacityReservation" }
  ]
```

PascalCase values consistent with PR #2025 ProgramType conventions.

}
```

### JSON Type Definition

``` json
{
  "definitions": {
    "commitmentProgramEntry": {
      "properties": {
        "ProgramType": { "type": "string" }
      }
    }
  },
  "properties": {
    "CommitmentPrograms": {
      "elements": { "ref": "commitmentProgramEntry" }
    }
  }
}
```

> NOTE: The above JSON Type Definition (JTD) is an approximation of the expected contents of this column, but it should not be considered normative because it cannot accurately describe the normative requirements (above) for CommitmentEligibilityDetails. Where there are discrepancies, preference will be given to the normative requirements.

Reviewer 5: ```suggestion
NOTE: The above JSON Type Definition (JTD) is an approximation of the expected contents of this column, but it should not be considered normative because it cannot accurately describe the normative requirements (above) for CommitmentProgramEligibility. Where there are discrepancies, preference will be given to the normative requirements.
```

## Example Scenarios

The examples below are not exhaustive and may change over time. Service providers are the authoritative source for their [*commitment programs*](#glossary:commitment-program).

### Provider A (Partially covered Virtual Machine Usage)

Scenario: A Virtual Machine usage row that is partially covered by Savings Plan. The eligibility column still reflects all programs this usage qualifies for, regardless of current coverage.

| ServiceProviderName | ServiceName      | CommitmentEligibilityDetails                                                                  |
|------------------|------------------|------------------|

> | Provider A          | Virtual Machines | {"CommitmentPrograms": [{"ProgramType": "SavingsPlan"}, {"ProgramType": "ReservedInstance"}]} |

Reviewer 5: ```suggestion
### Aura Web (Partially Covered Compute Usage)

Scenario: A compute usage row that is partially covered by a Flexible Spend Plan. The eligibility column still reflects all programs this usage qualifies for, regardless of current coverage.

| ServiceProviderName | ServiceName | CommitmentProgramEligibility                                                                            |
|---------------------|-------------|---------------------------------------------------------------------------------------------------------|
| Aura Web            | Compute     | {"CommitmentPrograms": [{"ProgramType": "FlexibleSpendPlan"}, {"ProgramType": "ResourceReservation"}]} |
```

Fictitious entity names from PR #2115.

### Provider B (Infrastructure with Subscription Commitment)

Scenario: An infrastructure host usage row eligible for Monthly and Annual commitment-based pricing, offering lower effective rates than On-Demand usage.

| ServiceProviderName | ServiceName    | CommitmentEligibilityDetails                                                                        |
|-----------|-----------|----------------------------------------|

> | Provider B          | Infrastructure | {"CommitmentPrograms": [{"ProgramType": "MonthlyCommitment"}, {"ProgramType": "AnnualCommitment"}]} |

Reviewer 5: ```suggestion
### StackLens (Observability with Platform Commitment)

Scenario: An observability platform usage row eligible for Monthly and Annual commitment-based pricing, offering lower effective rates than standard usage.

| ServiceProviderName | ServiceName   | CommitmentProgramEligibility                                                                                            |
|---------------------|---------------|-------------------------------------------------------------------------------------------------------------------------|
| StackLens           | Observability | {"CommitmentPrograms": [{"ProgramType": "MonthlyPlatformCommitment"}, {"ProgramType": "AnnualPlatformCommitment"}]} |
```

### Provider C (Ineligible Object Storage Usage)

Scenario: Standard object storage usage or a support fee, which is not eligible for any commitment program.

| ServiceProviderName | ServiceName   | CommitmentEligibilityDetails |
|---------------------|---------------|------------------------------|

> | Provider C          | ObjectStorage | null                         |

Reviewer 5: ```suggestion
### LatticeScale (Ineligible Object Storage Usage)

Scenario: Standard object storage usage or a support fee, which is not eligible for any commitment program.

| ServiceProviderName | ServiceName   | CommitmentProgramEligibility |
|---------------------|---------------|------------------------------|
| LatticeScale        | ObjectStorage | null                         |
```

### Provider D (Capacity Reservation-eligible Compute Usage)

A compute instance type and tenancy that are eligible for both Savings Plans/Reserved Instances and for capacity reservations (e.g., regional reservations, zonal reservations). The eligibility column reflects all commitment constructs the usage qualifies for.

| ServiceProviderName | ServiceName    | CommitmentEligibilityDetails                                                                                                                                               |
|------------------|------------------|------------------|

> | Provider D          | Compute | {"CommitmentPrograms": [{"ProgramType": "SavingsPlan"}, {"ProgramType": "ReservedInstance"}, {"ProgramType": "CapacityReservation"}, {"ProgramType": "ZonalReservation"}]} |

Reviewer 5: ```suggestion
### Aura Web (Capacity Reservation-Eligible Compute Usage)

Scenario: A compute instance type and tenancy that are eligible for both discount-bearing programs and capacity reservations. The eligibility column reflects all commitment constructs the usage qualifies for.

| ServiceProviderName | ServiceName | CommitmentProgramEligibility                                                                                                                                                                           |
|---------------------|-------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Aura Web            | Compute     | {"CommitmentPrograms": [{"ProgramType": "FlexibleSpendPlan"}, {"ProgramType": "ResourceReservation"}, {"ProgramType": "AdvanceCapacityReservation"}, {"ProgramType": "ZonalCapacityReservation"}]} |
```

### Object ID

CommitmentEligibilityDetailsObject

### Object Display Name

> Commitment Eligibility Details Object

Reviewer 5: ```suggestion
### Object ID

CommitmentProgramEligibilityObject

### Object Display Name

Commitment Program Eligibility Object
```

## Column ID

CommitmentEligibilityDetails

## Display Name

> Commitment Eligibility Details

Reviewer 5: ```suggestion
## Column ID

CommitmentProgramEligibility

## Display Name

Commitment Program Eligibility
```

## Content Constraints

| Constraint    | Value                                                                                                                        |
|:-------------------------------------|:---------------------------------|
| Dataset       | [Cost and Usage](#datasets.costandusage)                                                                                     |
| Column type   | Dimension                                                                                                                    |
| Feature level | Conditional                                                                                                                  |
| Allows nulls  | True                                                                                                                         |
| Data type     | JSON                                                                                                                         |
| Value format  | [JsonObjectFormat](#attributes.jsonobjectformat)                                                                             |

> | Object        | [CommitmentEligibilityDetailsObject](#datasets.costandusage.commitmenteligibilitydetails.commitmenteligibilitydetailsobject) |

Reviewer 5: ```suggestion
| Object        | [CommitmentProgramEligibilityObject](#datasets.costandusage.commitmentprogrameligibility.commitmentprogrameligibilityobject) |
```
